11-Backbonejs/06-Adding-Custom-Events.md

```diff
@@ -0,0 +1,257 @@
+# Adding Custom Events
+
+## Learning Goals
+By the end of this document you should be able:
+-   to explain how the Backbone Event API works
+-   to add custom events to Backbone Views, Collections & Models.
+-   to explain the concept of an Event Bus
+-   to use an Event Bus to enable Backbone objects to communicate.  
+
+## Adding An Event To An Object
+
+In Backbone Models contain information and business logic and Views render model data.  Events are how Views & Models communicate.  
+
+You've seen jQuery Events like below:
+
+```html
+<button id="sample-button">Press Me</button>
+
+<script>
+$('#sample-button').on('click', function() {
+  alert("You Pressed it");
+})
+</script>
+```
+
+In this we added a jQuery Event to the button and caused an alert to 'fire' when the button was pressed.  Similarly we can use the Backbone.Events module to give any object the ability to register and 'trigger' events.  
+
+Below we have created a generic JavaScript object and then used Underscore to mix-in the Backbone.Events module into the object.  Then we add an event called, "Past_Due" to the object and Triggered the event with the `.trigger` function.  
+
+Notice the `_.extend(task, Backbone.Events)` line.  The authors of Backbone have created a way to define custom events with an Object called Events, similar to the DOM events we have used earlier.  Events provides a few functions (`on()` and `listenTo()`) for event handlers to subscribe to events and a single function to "trigger" an event (named `trigger`).  
+
+Remember [Ruby Mix-ins](https://www.tutorialspoint.com/ruby/ruby_modules.htm)?  In Ruby we can define a model and use include or extend to bring those methods into existing classes as static or instance methods.  By default JavaScript doesn't really do Mix-ins, but the wonderful developers of Underscore provides a way to **Mix-in** functions from one object into another.  So below we Rubyfy Javascript to add in the Mixin 
+
+```javascript
+// generic Javascript example
+
+var task = {
+  assignee: "Dan",
+  task:     "Do work",
+  due_date: "01/01/2017"
+};
+
+_.extend(task, Backbone.Events);
+
+task.on("PastDue", function(message) {
```

Comment:
We're listening for a `PastDue` event here, but triggering a `Past_Due` event. Also further on in  the paragraph under "Events: on, off & trigger" we use an event named `pastDue`.